sql-to-dbsp-compiler/lib/json/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+#![deny(missing_docs)]
+
+//! Functions related to `JSON` support in DBSP
+
+/// The actual Json type.
+pub type Json = ijson::IValue;
+
+/// Parses the given string to [`Json`].
+///
+/// # Panics
+/// - Panics if it fails to deserialize to a valid Json.
+pub fn parse_json(value: impl AsRef<str>) -> Json {
+    serde_json::from_str(value.as_ref()).unwrap_or_else(|_| {
+        panic!(
+            "cannot deserialize given string to JSON: {}",
+            value.as_ref()
+        )
+    })
+}
+
+/// Returns `true` if the given string is a valid [`Json`].
+pub fn check_json(value: impl AsRef<str>) -> bool {
+    serde_json::from_str::<Json>(value.as_ref()).is_ok()
+}
+
+/// Extracts a `field` from the [`Json`] value.
+/// Use [`try_json_field`] for a non panicking implementation.
+///
+/// # Panics
+/// - Panics if the JSON value isn't a JSON object.
+/// - Panics if the `field` doesn't exist in this object.
+pub fn json_field(value: Json, field: impl AsRef<str>) -> Json {
+    value
+        .get(field.as_ref())
+        .expect("invalid: JSON_FIELD called with non existent field")
+        .clone()
+}
+
+/// Tries to extract a `field` from the [`Json`] value.
+pub fn try_json_field(value: Json, field: impl AsRef<str>) -> Option<Json> {
+    value.as_object()?.get(field.as_ref()).cloned()
+}
+
+/// Extracts the value at the given JSON index from this JSON array literal.
+/// Indexing starts from 1 to match other SQL functions.
+/// Use [`try_json_index`] for a non panicking implementation.
+///
+/// # Panics
+/// - Panics if the JSON value isn't a JSON array literal.
+/// - Panics if called with the index `0`.
+pub fn json_index(value: Json, mut idx: usize) -> Json {
+    idx = idx
+        .checked_sub(1)
+        .expect("invalid: JSON_INDEX called with index 0");
+
+    value
+        .get(idx)
+        .expect("invalid: no value in the given index")
+        .clone()
+}
+
+/// Tries to extract the value at the given index from this JSON array literal.
+pub fn try_json_index(value: Json, mut idx: usize) -> Option<Json> {
+    idx = idx.checked_sub(1)?;
+
+    value.as_array()?.get(idx).cloned()
+}
+
+/// Deserialize this JSON `value` as the given type: `T`.
+/// `T` must have implement [`serde::de::DeserializeOwned`]
+///
+/// # Panics
+/// - Panics if deserializing to the given type `T` fails.
+pub fn json_as<T: serde::de::DeserializeOwned>(value: Json) -> T {
+    ijson::from_value(&value).expect("failed to deserialize JSON to the given type")
+}
+
+/// Serialize this JSON `value` as a string.
+///
+/// # Panics
+/// - Panics if serialization fails.
+pub fn to_string(value: Json) -> String {
+    serde_json::to_string(&value).expect("failed to cast JSON to string")
```

Comment:
I am not sure "cast" is the right word. 
Perhaps "reinterpret"?

Reply:
how about the other primitive types supported by JSON? boolean, decimal, perhaps fp, int too?

Reply:
For them, `json_as::<f64>(json)` would work. 
Although, right now, we only support casting to `f64` and not `Decimal`.

Reply:
> I am not sure "cast" is the right word.
> Perhaps "reinterpret"?

Maybe we should just say "serialize".

Reply:
no, it's not serialize, that always succeeds on a JSON value.
This would only succeed if the JSON value *is* a string like `"string"`

Reply:
Actually, this function literally serializes the JSON to `String`. 

If we want it to extract the String literal value from JSON, we'd need something else.

Reply:
we need both.
this one you can call "serialize'

Reply:
To only extract the string literal if the type is String, we can do: `json_as::<String>(json)`.